text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+
+### Yanking gems
+
+The hashed filenames pose a slight challenge when yanking a gem.
+The relative infrequency of yanking means that we can accept a more involved process removing files than adding them or hosting them.
+
+When a version is yanked, the file manifest for that version will be used to search for any hashed files that are unique to the yanked gem.
+If a SHA256 named file would be orphaned by yanking the gem (it has no gem versions that reference it) then the file will be deleted from S3 and purged from Fastly's cache.
+If a file has other references, then the file was already a duplicate of a file in a previous version. Files with multiple references will be preserved until the last reference is removed.
+
+When a gem is yanked the manifest table rows recording the file names with checksum, size could be preserved but no longer shown anywhere.
+
+### Indexing older gem versions
+
+Depending on the implementation of future fetaures, gems can be indexed on demand by request or in an automated way.
+
+### Impact
+
+The gem ingestion process should be invisible to end users.
+Once the data is used to serve an API, users can expect a slight delay before a newly uploaded gem is available in the content API.
+Yanking a gem might also have a slight delay before files are not available at the checksum urls, but this should be expected already.
+
+Future features and external apps that use the contents API will have a reliable and authoritative source for gem contents that can be accessed on demand.
+
+# Reference-level explanation
+
+### Ingestion process
+
+The following process will happen when a gem is uploaded or manually triggered:
+
+1. A background job is enqueued with the gem version.
+2. The worker pulls down the gem from S3.
+3. The gem in expanded into a temporary directory.
+4. Each file is enumerated, skipping directories.
+5. The worker uses a tool (grep is good at this) to mark all files that contain binary data.
+6. Textual files (not binaries) will have lines counted.
+7. All files will have size, content_type and SHA256 checksums recorded.
+8. A record is created for each file, recording the attributes:
+  ```
+  version_id
+  path (relative to the gem root)
+  size
+  binary (boolean)
+  content_type (mime type determined by `file` command or libmagic)
+  sha256 (checksum)
+  created_at
+  yanked_at (optional, see unresolved questions)
+  ```
+9. New text files are uploaded to a gem contents s3 bucket with key: `/(gemname)/files/(sha256)`
+  (The content S3 bucket is checked to see if a file already exists at the key before uploading.)
+10. An attribute is updated on the Version record to indicate the gem content has been successfully uploaded.
+  (This is necessary because a partially ingested gem would show a partial file list. Better to show all or nothing.)
+
+### Yank process
+
+When a gem is yanked, additional steps will be performed to remove files and stop serving manifests.
```

Comment:
we will likely need to also update the fastly purging in the yank codepath to account for the new files being deleted

Reply:
I'll add mention of that. Thank you!